File: src/main/java/client/messages/broadcast/types/SpeakerworldMessage.java
package client.messages.broadcast.types;

import client.messages.broadcast.AbstractBroadcastMessage;
import client.messages.broadcast.BroadcastMessageType;
import util.packet.PacketWriter;

/**
 * Color: Purple on pink bg
 * Prefix: Medalname charactername ch01
 * <p>
 * Super Megaphone
 */
public class SpeakerworldMessage extends AbstractBroadcastMessage {

    private final String message;
    private final String medal;
    private final String name;
    private final byte channel;
    private final boolean ear;

    public SpeakerworldMessage(String message, String medal, String name, int channel, boolean ear) {
        this.message = message;
        this.medal = medal;
        this.name = name;
        this.channel = (byte) channel;
        this.ear = ear;
    }

    @Override
    public BroadcastMessageType getType() {
        return BroadcastMessageType.SPEAKERWORLD;
    }

    @Override
    protected void encodeData(PacketWriter pw) {
        pw.writeMapleString(medal + name + " : " + message);
        pw.write(channel);
        pw.writeBool(ear);
    }
}

File: src/main/java/client/messages/broadcast/types/SpeakerworldMessage.kt
package client.messages.broadcast.types

import client.messages.broadcast.AbstractBroadcastMessage
import client.messages.broadcast.BroadcastMessageType
import util.packet.PacketWriter

/**
 * Color: Purple on pink bg
 * Prefix: Medalname charactername ch01
 *
 * Super Megaphone
 */
class SpeakerworldMessage(private val message: String, private val medal: String, private val name: String, channel: Int, private val ear: Boolean) : AbstractBroadcastMessage() {

    private val channel: Byte = channel.toByte()

    override val type: BroadcastMessageType get() = BroadcastMessageType.SPEAKERWORLD

    override fun encodeData(pw: PacketWriter) {
        pw.writeMapleString("$medal$name : $message")
        pw.write(channel.toInt())
        pw.writeBool(ear)
    }
}